File: twin/toy_cascade.py
"""The explainer's low-complexity DC screening example, solved on the server.

The explainer page (`web/src/pages/ExplainerPage.tsx`) teaches the chain a DC
screening model follows: balance island injections, solve line flows, compare a
flow to a rating, trip the worst overload, solve again. It used to run that
chain in the browser, which put a second simulation engine next to
`twin/cascade.py` and made the browser compute an answer. This module owns the
teaching solve instead; `scripts/export_toy_cascade_trace.py` freezes its output
into `data/explainer/toy-cascade-trace.json`, and the page replays that trace.

This is deliberately NOT `twin/cascade.py`. It is a five-bus teaching network
with hand-picked reactances and ratings, no pandapower, no DuckDB, no weather
and no county attribution. It is not the product's solver and must not be
reused as one; see `docs/specs/03-cascade-sim.md` -> "Explainer teaching solve".
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from typing import Any, Literal

SCHEMA_VERSION = 1

#: The initiating outage the lesson is built around.
INITIATING_LINE_ID = "hub-east"

#: The teaching rule trips at most this many lines after the initiating outage.
MAX_CASCADE_TRIPS = 3

_ZERO_TOLERANCE = 1e-9


@dataclass(frozen=True)
class ToyBus:
    id: str
    name: str
    generation_mw: float
    demand_mw: float
    x: float
    y: float


@dataclass(frozen=True)
class ToyLine:
    id: str
    from_bus: str
    to_bus: str
    reactance: float
    rating_mw: float


class ToyCascadeError(RuntimeError):
    """A named refusal from the teaching solve. Never a plausible default."""


TOY_BUSES: tuple[ToyBus, ...] = (
    ToyBus("west", "West generator", 120.0, 0.0, 84, 185),
    ToyBus("north", "North load", 40.0, 50.0, 255, 65),
    ToyBus("hub", "Central hub", 0.0, 30.0, 390, 185),
    ToyBus("east", "East load", 0.0, 70.0, 590, 125),
    ToyBus("south", "South load", 0.0, 10.0, 490, 330),
)

TOY_LINES: tuple[ToyLine, ...] = (
    ToyLine("west-north", "west", "north", 0.25, 80.0),
    ToyLine("west-hub", "west", "hub", 0.20, 110.0),
    ToyLine("north-hub", "north", "hub", 0.25, 80.0),
    ToyLine("hub-east", "hub", "east", 0.20, 90.0),
    ToyLine("hub-south", "hub", "south", 0.25, 70.0),
    ToyLine("east-south", "east", "south", 0.25, 35.0),
)

_BUS_BY_ID = {bus.id: bus for bus in TOY_BUSES}
_LINE_BY_ID = {line.id: line for line in TOY_LINES}


def _number(value: float) -> float:
    """Collapse solver noise to a signed-zero-free 0 and fix to 6 decimals."""
    return 0.0 if abs(value) < _ZERO_TOLERANCE else round(value, 6) + 0.0


def _connected_components(active_lines: tuple[ToyLine, ...]) -> list[list[str]]:
    neighbours: dict[str, list[str]] = {bus.id: [] for bus in TOY_BUSES}
    for line in active_lines:
        neighbours[line.from_bus].append(line.to_bus)
        neighbours[line.to_bus].append(line.from_bus)
    unseen = [bus.id for bus in TOY_BUSES]
    components: list[list[str]] = []
    while unseen:
        first = unseen.pop(0)
        component = [first]
        todo = [first]
        while todo:
            current = todo.pop()
            for neighbour in neighbours[current]:
                if neighbour in unseen:
                    unseen.remove(neighbour)
                    component.append(neighbour)
                    todo.append(neighbour)
        components.append(component)
    return components


def _balance_components(
    active_lines: tuple[ToyLine, ...],
) -> tuple[dict[str, float], list[dict[str, Any]]]:
    injections = {bus.id: bus.generation_mw - bus.demand_mw for bus in TOY_BUSES}
    actions: list[dict[str, Any]] = []
    for component in _connected_components(active_lines):
        total = sum(injections[bus_id] for bus_id in component)
        if abs(total) < _ZERO_TOLERANCE:
            continue
        shedding = total < 0
        candidates = [
            _BUS_BY_ID[bus_id]
            for bus_id in component
            if (
                _BUS_BY_ID[bus_id].demand_mw
                if shedding
                else _BUS_BY_ID[bus_id].generation_mw
            )
            > 0
        ]
        denominator = sum(
            (bus.demand_mw if shedding else bus.generation_mw) for bus in candidates
        )
        if not denominator:
            raise ToyCascadeError(
                f"Cannot balance component containing {', '.join(component)}."
            )
        for bus in candidates:
            share = bus.demand_mw if shedding else bus.generation_mw
            mw = abs(total) * (share / denominator)
            injections[bus.id] += mw if shedding else -mw
            actions.append(
                {
                    "busId": bus.id,
                    "kind": "shed_load" if shedding else "curtail_generation",
                    "mw": _number(mw),
                }
            )
    return injections, actions


def _solve_linear_system(matrix: list[list[float]], vector: list[float]) -> list[float]:
    """Gauss-Jordan elimination. The teaching system is at most 4x4."""
    size = len(matrix)
    augmented = [list(row) + [vector[index]] for index, row in enumerate(matrix)]
    for column in range(size):
        pivot = max(range(column, size), key=lambda row: abs(augmented[row][column]))
        if abs(augmented[pivot][column]) < _ZERO_TOLERANCE:
            raise ToyCascadeError("Toy DC matrix is singular.")
        augmented[column], augmented[pivot] = augmented[pivot], augmented[column]
        divisor = augmented[column][column]
        augmented[column] = [value / divisor for value in augmented[column]]
        for row in range(size):
            if row == column:
                continue
            factor = augmented[row][column]
            augmented[row] = [
                value - factor * augmented[column][index]
                for index, value in enumerate(augmented[row])
            ]
    return [_number(row[-1]) for row in augmented]


def solve_toy_dc(active_line_ids: frozenset[str]) -> dict[str, Any]:
    """Balance every island, then solve B-theta for the active corridors."""
    active_lines = tuple(line for line in TOY_LINES if line.id in active_line_ids)
    injections, actions = _balance_components(active_lines)
    angles = {bus.id: 0.0 for bus in TOY_BUSES}
    for component in _connected_components(active_lines):
        if len(component) < 2:
            continue
        unknowns = [bus_id for bus_id in component if bus_id != component[0]]
        index = {bus_id: position for position, bus_id in enumerate(unknowns)}
        matrix = [[0.0] * len(unknowns) for _ in unknowns]
        vector = [injections[bus_id] for bus_id in unknowns]
        for line in active_lines:
            susceptance = 1 / line.reactance
            from_index = index.get(line.from_bus)
            to_index = index.get(line.to_bus)
            if from_index is not None:
                matrix[from_index][from_index] += susceptance
                if to_index is not None:
                    matrix[from_index][to_index] -= susceptance
            if to_index is not None:
                matrix[to_index][to_index] += susceptance
                if from_index is not None:
                    matrix[to_index][from_index] -= susceptance
        solution = _solve_linear_system(matrix, vector)
        for bus_id in unknowns:
            angles[bus_id] = solution[index[bus_id]]
    lines = []
    for line in active_lines:
        flow_mw = _number(
            (angles[line.from_bus] - angles[line.to_bus]) / line.reactance
        )
        lines.append(
            {
                "id": line.id,
                "from": line.from_bus,
                "to": line.to_bus,
                "reactance": line.reactance,
                "ratingMw": line.rating_mw,
                "flowMw": flow_mw,
                "utilizationPct": _number(abs(flow_mw) / line.rating_mw * 100),
            }
        )
    return {
        "injections": {bus_id: _number(value) for bus_id, value in injections.items()},
        "actions": actions,
        "angles": angles,
        "lines": lines,
    }


def _most_overloaded(lines: list[dict[str, Any]]) -> str | None:
    overloaded = [line for line in lines if line["utilizationPct"] > 100]
    if not overloaded:
        return None
    overloaded.sort(key=lambda line: (-line["utilizationPct"], line["id"]))
    return str(overloaded[0]["id"])


def run_toy_cascade() -> list[dict[str, Any]]:
    """The seeded outage, the re-dispatch, and the overload trips that follow."""
    open_lines: set[str] = set()
    stages: list[dict[str, Any]] = []

    def add_stage(
        stage_id: str, title: str, explanation: str, tripped_line_id: str | None
    ) -> None:
        active_line_ids = [line.id for line in TOY_LINES if line.id not in open_lines]
        solved = solve_toy_dc(frozenset(active_line_ids))
        stages.append(
            {
                "id": stage_id,
                "title": title,
                "explanation": explanation,
                "trippedLineId": tripped_line_id,
                "activeLineIds": active_line_ids,
                "injectionsMw": solved["injections"],
                "angles": solved["angles"],
                "balanceActions": solved["actions"],
                "lines": solved["lines"],
                "nextTripLineId": _most_overloaded(solved["lines"]),
            }
        )

    add_stage(
        "base",
        "1. Normal toy network",
        "All six synthetic corridors are available. DC flow balances the five specified bus injections.",
        None,
    )
    open_lines.add(INITIATING_LINE_ID)
    add_stage(
        "event",
        "2. Synthetic initiating outage",
        "For teaching purposes, Central hub → East load trips. Re-solving redistributes its power through East load → South load.",
        INITIATING_LINE_ID,
    )
    for step in range(MAX_CASCADE_TRIPS):
        next_trip = stages[-1]["nextTripLineId"]
        if not next_trip:
            break
        open_lines.add(next_trip)
        line = _LINE_BY_ID[next_trip]
        add_stage(
            f"cascade-{step + 1}",
            f"3. Cascade trip {step + 1}",
            f"{line.from_bus} → {line.to_bus} exceeds its thermal rating, so this toy rule removes the most overloaded remaining line and re-solves.",
            next_trip,
        )
    return stages


def trace_hash(stages: list[dict[str, Any]]) -> str:
    """A stable digest of the frozen stages, so a replayed trace is checkable."""
    canonical = json.dumps(stages, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def toy_cascade_trace() -> dict[str, Any]:
    """The whole persisted artifact: network, stages, and the digest over them."""
    stages = run_toy_cascade()
    kind: Literal["synthetic_five_bus_teaching_network"] = (
        "synthetic_five_bus_teaching_network"
    )
    return {
        "schemaVersion": SCHEMA_VERSION,
        "modelFidelity": "dc_screening",
        "networkProvenance": kind,
        "networkLabel": "synthetic five-bus teaching network",
        "limitations": [
            "Synthetic five-bus teaching network; not ACTIVSg2000, not a physical asset, and not an interconnection result.",
            "DC screening excludes AC voltage, transient stability, protection, unit commitment, and regulatory feasibility.",
            "This teaching solve is deliberately independent of twin/cascade.py and is not the product's solver.",
        ],
        "network": {
            "buses": [
                {
                    "id": bus.id,
                    "name": bus.name,
                    "generationMw": bus.generation_mw,
                    "demandMw": bus.demand_mw,
                    "x": bus.x,
                    "y": bus.y,
                }
                for bus in TOY_BUSES
            ],
            "lines": [
                {
                    "id": line.id,
                    "from": line.from_bus,
                    "to": line.to_bus,
                    "reactance": line.reactance,
                    "ratingMw": line.rating_mw,
                }
                for line in TOY_LINES
            ],
        },
        "stages": stages,
        "traceHash": trace_hash(stages),
    }

File: web/src/pages/toyCascade.ts
/**
 * A deliberately small DC-screening example for the explainer page.
 *
 * This is synthetic teaching data, separate from the main-page fixture. The
 * values make one initiating outage visibly reroute power through an
 * overloaded corridor before an islanded load is shed. It is not a grid study,
 * forecast, or protection model.
 */
export interface ToyBus {
  readonly id: string;
  readonly name: string;
  readonly generationMw: number;
  readonly demandMw: number;
  readonly x: number;
  readonly y: number;
}

export interface ToyLine {
  readonly id: string;
  readonly from: string;
  readonly to: string;
  readonly reactance: number;
  readonly ratingMw: number;
}

export interface BalanceAction {
  readonly busId: string;
  readonly kind: "shed_load" | "curtail_generation";
  readonly mw: number;
}

export interface SolvedLine extends ToyLine {
  readonly flowMw: number;
  readonly utilizationPct: number;
}

export interface CascadeStage {
  readonly id: string;
  readonly title: string;
  readonly explanation: string;
  readonly trippedLineId: string | null;
  readonly activeLineIds: readonly string[];
  readonly injectionsMw: Readonly<Record<string, number>>;
  readonly angles: Readonly<Record<string, number>>;
  readonly balanceActions: readonly BalanceAction[];
  readonly lines: readonly SolvedLine[];
  readonly nextTripLineId: string | null;
}

export const TOY_BUSES: readonly ToyBus[] = [
  { id: "west", name: "West generator", generationMw: 120, demandMw: 0, x: 84, y: 185 },
  { id: "north", name: "North load", generationMw: 40, demandMw: 50, x: 255, y: 65 },
  { id: "hub", name: "Central hub", generationMw: 0, demandMw: 30, x: 390, y: 185 },
  { id: "east", name: "East load", generationMw: 0, demandMw: 70, x: 590, y: 125 },
  { id: "south", name: "South load", generationMw: 0, demandMw: 10, x: 490, y: 330 },
] as const;

export const TOY_LINES: readonly ToyLine[] = [
  { id: "west-north", from: "west", to: "north", reactance: 0.25, ratingMw: 80 },
  { id: "west-hub", from: "west", to: "hub", reactance: 0.2, ratingMw: 110 },
  { id: "north-hub", from: "north", to: "hub", reactance: 0.25, ratingMw: 80 },
  { id: "hub-east", from: "hub", to: "east", reactance: 0.2, ratingMw: 90 },
  { id: "hub-south", from: "hub", to: "south", reactance: 0.25, ratingMw: 70 },
  { id: "east-south", from: "east", to: "south", reactance: 0.25, ratingMw: 35 },
] as const;

const BUS_BY_ID = new Map(TOY_BUSES.map((bus) => [bus.id, bus]));
const LINE_BY_ID = new Map(TOY_LINES.map((line) => [line.id, line]));

function number(value: number): number {
  return Math.abs(value) < 1e-9 ? 0 : Number(value.toFixed(6));
}

function connectedComponents(activeLines: readonly ToyLine[]): string[][] {
  const neighbors = new Map(TOY_BUSES.map((bus) => [bus.id, [] as string[]]));
  for (const line of activeLines) {
    neighbors.get(line.from)?.push(line.to);
    neighbors.get(line.to)?.push(line.from);
  }
  const unseen = new Set(TOY_BUSES.map((bus) => bus.id));
  const components: string[][] = [];
  while (unseen.size) {
    const first = unseen.values().next().value as string;
    const component: string[] = [];
    const todo = [first];
    unseen.delete(first);
    while (todo.length) {
      const id = todo.pop()!;
      component.push(id);
      for (const neighbor of neighbors.get(id) ?? []) if (unseen.delete(neighbor)) todo.push(neighbor);
    }
    components.push(component);
  }
  return components;
}

function balanceComponents(activeLines: readonly ToyLine[]) {
  const injections: Record<string, number> = Object.fromEntries(
    TOY_BUSES.map((bus) => [bus.id, bus.generationMw - bus.demandMw]),
  );
  const actions: BalanceAction[] = [];
  for (const component of connectedComponents(activeLines)) {
    const total = component.reduce((sum, id) => sum + injections[id], 0);
    if (Math.abs(total) < 1e-9) continue;
    const candidates = component.map((id) => BUS_BY_ID.get(id)!).filter((bus) => total < 0 ? bus.demandMw > 0 : bus.generationMw > 0);
    const denominator = candidates.reduce((sum, bus) => sum + (total < 0 ? bus.demandMw : bus.generationMw), 0);
    if (!denominator) throw new Error(`Cannot balance component containing ${component.join(", ")}.`);
    for (const bus of candidates) {
      const mw = Math.abs(total) * ((total < 0 ? bus.demandMw : bus.generationMw) / denominator);
      injections[bus.id] += total < 0 ? mw : -mw;
      actions.push({ busId: bus.id, kind: total < 0 ? "shed_load" : "curtail_generation", mw: number(mw) });
    }
  }
  return { injections, actions };
}

/** Solve A·x=b with Gauss-Jordan elimination. The toy system is at most 4×4. */
function solveLinearSystem(matrix: number[][], vector: number[]): number[] {
  const augmented = matrix.map((row, index) => [...row, vector[index]]);
  for (let column = 0; column < matrix.length; column += 1) {
    let pivot = column;
    for (let row = column + 1; row < matrix.length; row += 1) if (Math.abs(augmented[row][column]) > Math.abs(augmented[pivot][column])) pivot = row;
    if (Math.abs(augmented[pivot][column]) < 1e-9) throw new Error("Toy DC matrix is singular.");
    [augmented[column], augmented[pivot]] = [augmented[pivot], augmented[column]];
    const divisor = augmented[column][column];
    augmented[column] = augmented[column].map((value) => value / divisor);
    for (let row = 0; row < augmented.length; row += 1) {
      if (row === column) continue;
      const factor = augmented[row][column];
      augmented[row] = augmented[row].map((value, index) => value - factor * augmented[column][index]);
    }
  }
  return augmented.map((row) => number(row[row.length - 1]));
}

export function solveToyDc(activeLineIds: ReadonlySet<string>) {
  const activeLines = TOY_LINES.filter((line) => activeLineIds.has(line.id));
  const { injections, actions } = balanceComponents(activeLines);
  const angles: Record<string, number> = Object.fromEntries(TOY_BUSES.map((bus) => [bus.id, 0]));
  for (const component of connectedComponents(activeLines)) {
    if (component.length < 2) continue;
    const reference = component[0];
    const unknowns = component.filter((id) => id !== reference);
    const index = new Map(unknowns.map((id, position) => [id, position]));
    const matrix = unknowns.map(() => unknowns.map(() => 0));
    const vector = unknowns.map((id) => injections[id]);
    for (const line of activeLines) {
      const susceptance = 1 / line.reactance;
      const fromIndex = index.get(line.from);
      const toIndex = index.get(line.to);
      if (fromIndex !== undefined) {
        matrix[fromIndex][fromIndex] += susceptance;
        if (toIndex !== undefined) matrix[fromIndex][toIndex] -= susceptance;
      }
      if (toIndex !== undefined) {
        matrix[toIndex][toIndex] += susceptance;
        if (fromIndex !== undefined) matrix[toIndex][fromIndex] -= susceptance;
      }
    }
    const solution = solveLinearSystem(matrix, vector);
    for (const id of unknowns) angles[id] = solution[index.get(id)!];
  }
  const lines = activeLines.map((line) => {
    const flowMw = number((angles[line.from] - angles[line.to]) / line.reactance);
    return { ...line, flowMw, utilizationPct: number((Math.abs(flowMw) / line.ratingMw) * 100) };
  });
  return { activeLines, injections, actions, angles, lines };
}

function mostOverloaded(lines: readonly SolvedLine[]): SolvedLine | null {
  return lines.filter((line) => line.utilizationPct > 100).sort((left, right) => right.utilizationPct - left.utilizationPct || left.id.localeCompare(right.id))[0] ?? null;
}

/** Run the same tiny chain shown in the page: a seeded outage, re-dispatch, then overload trip(s). */
export function runToyCascade(): readonly CascadeStage[] {
  const open = new Set<string>();
  const stages: CascadeStage[] = [];
  const addStage = (id: string, title: string, explanation: string, trippedLineId: string | null) => {
    const activeLineIds = TOY_LINES.filter((line) => !open.has(line.id)).map((line) => line.id);
    const solved = solveToyDc(new Set(activeLineIds));
    const nextTripLineId = mostOverloaded(solved.lines)?.id ?? null;
    stages.push({
      id, title, explanation, trippedLineId, activeLineIds, nextTripLineId,
      injectionsMw: solved.injections,
      angles: solved.angles,
      balanceActions: solved.actions,
      lines: solved.lines,
    });
  };
  addStage("base", "1. Normal toy network", "All six synthetic corridors are available. DC flow balances the five specified bus injections.", null);
  const initiating = "hub-east";
  open.add(initiating);
  addStage("event", "2. Synthetic initiating outage", "For teaching purposes, Central hub → East load trips. Re-solving redistributes its power through East load → South load.", initiating);
  for (let index = 0; index < 3; index += 1) {
    const next = stages[stages.length - 1].nextTripLineId;
    if (!next) break;
    open.add(next);
    const line = LINE_BY_ID.get(next)!;
    addStage(`cascade-${index + 1}`, `3. Cascade trip ${index + 1}`, `${line.from} → ${line.to} exceeds its thermal rating, so this toy rule removes the most overloaded remaining line and re-solves.`, next);
  }
  return stages;
}
